include/deal.II/differentiation/ad/ad_drivers.h

```diff
@@ -0,0 +1,1124 @@
+
+    // Specialization for taped Adol-C auto-differentiable numbers.
+    //
+    // Note: In the case of Adol-C taped numbers, the associated scalar
+    // type is always expected to be a double. So we need to make a further
+    // specialization when ScalarType is a float.
+    template<typename ADNumberType>
+    struct TapedDrivers<ADNumberType,double,typename std::enable_if<
+      ADNumberTraits<ADNumberType>::type_code == NumberTypes::adolc_taped
+      >::type>
+    {
+      typedef double scalar_type;
+
+      // === Taping ===
+
+      static void
+      enable_taping(const unsigned int &tape_index,
+                    const bool         &keep)
+      {
+        trace_on(tape_index,keep);
+      }
+
+      static void
+      disable_taping(const unsigned int &active_tape_index,
+                     const bool         &write_tapes_to_file)
+      {
+        if (write_tapes_to_file)
+          {
+            trace_off(active_tape_index); // Slow
+
+            std::vector<std::size_t> counts (STAT_SIZE);
+#ifdef DEAL_II_WITH_ADOLC
+            ::tapestats(active_tape_index, counts.data());
+#else
+            AssertThrow(false, ExcNotImplemented());
+#endif
+          }
+        else
+          trace_off(); // Fast(er)
+      }
+
+      // === Scalar drivers ===
+
+      static scalar_type
+      value (const unsigned int             &active_tape_index,
+             const unsigned int             &n_independent_variables,
+             const std::vector<scalar_type> &independent_variables)
+      {
+        Assert(independent_variables.size() == n_independent_variables,
+               ExcDimensionMismatch(independent_variables.size(),n_independent_variables));
+
+        double *f = new double();
+
+#ifdef DEAL_II_WITH_ADOLC
+        ::function(active_tape_index,
+                   1, // Only one dependent variable
+                   n_independent_variables,
+                   const_cast<double *>(independent_variables.data()),
+                   f);
+#else
+        AssertThrow(false, ExcNotImplemented());
+#endif
+
+        const double value = f[0];
+
+        // Cleanup :-/
+        delete f;
+        f = nullptr;
+
+        return value;
+      }
+
+      static void
+      gradient (const unsigned int             &active_tape_index,
+                const unsigned int             &n_independent_variables,
+                const std::vector<scalar_type> &independent_variables,
+                Vector<scalar_type>            &gradient)
+      {
+        Assert(independent_variables.size() == n_independent_variables,
+               ExcDimensionMismatch(independent_variables.size(),n_independent_variables));
+        Assert(gradient.size() == n_independent_variables,
+               ExcMessage("The length of the gradient vector must be equal "
+                          "to the number of independent variables."));
+
+        scalar_type *g = new scalar_type[n_independent_variables]; // Use smart pointers or std_cxx11::array?
```

Comment:
`std::vector< scalar_type> g(n_independent_variables)`? Then give `g.begin()` or something to adolc?

Reply:
p.s. `std::array` need compile-time size, which is not the case here.

Reply:
Yes, the note was an old todo. `std::array` definitely wouldn't work here.